packages/x-tree-view-pro/src/internals/plugins/useTreeViewItemsReordering/useTreeViewItemsReordering.types.ts

```diff
@@ -0,0 +1,157 @@
+   * @returns {boolean} `true` if the item can be reordered.
+   * @default () => true
+   */
+  isItemReorderable?: (itemId: string) => boolean;
+  /**
+   * Used to determine if a given item can move to some new position.
+   * @param {object} params The params describing the item re-ordering.
+   * @param {string} params.itemId The id of the item to check.
+   * @param {TreeViewItemReorderPosition} params.oldPosition The old position of the item.
+   * @param {TreeViewItemReorderPosition} params.newPosition The new position of the item.
+   * @returns {boolean} `true` if the item can move to the new position.
+   */
+  canMoveItemToNewPosition?: (params: {
+    itemId: string;
+    oldPosition: TreeViewItemReorderPosition;
+    newPosition: TreeViewItemReorderPosition;
+  }) => boolean;
+  /**
+   * Callback fired when a tree item is moved in the tree.
+   * @param {object} params The params describing the item re-ordering.
+   * @param {string} params.itemId The id of the item moved.
+   * @param {TreeViewItemReorderPosition} params.oldPosition The old position of the item.
+   * @param {TreeViewItemReorderPosition} params.newPosition The new position of the item.
+   */
+  onItemPositionChange?: (params: {
+    itemId: string;
+    oldPosition: TreeViewItemReorderPosition;
+    newPosition: TreeViewItemReorderPosition;
+  }) => void;
+}
+
+export type UseTreeViewItemsReorderingDefaultizedParameters = DefaultizedProps<
+  UseTreeViewItemsReorderingParameters,
+  'itemsReordering'
+>;
+
+export interface UseTreeViewItemsReorderingState {
+  itemsReordering: {
+    draggedItemId: string;
+    targetItemId: string;
+    newPosition: TreeViewItemReorderPosition | null;
+    action: TreeViewItemsReorderingAction | null;
+  } | null;
+}
+
+interface UseTreeViewItemsReorderingContextValue {
+  itemsReordering: {
+    enabled: boolean;
+    currentDrag: UseTreeViewItemsReorderingState['itemsReordering'];
+  };
+}
+
+export type UseTreeViewItemsReorderingSignature = TreeViewPluginSignature<{
+  params: UseTreeViewItemsReorderingParameters;
+  defaultizedParams: UseTreeViewItemsReorderingDefaultizedParameters;
+  instance: UseTreeViewItemsReorderingInstance;
+  state: UseTreeViewItemsReorderingState;
+  contextValue: UseTreeViewItemsReorderingContextValue;
+  dependantPlugins: [UseTreeViewItemsSignature];
+}>;
+
+export interface UseTreeItem2RootSlotPropsFromItemsReordering {
+  draggable?: true;
+  onDragStart?: MuiCancellableEventHandler<React.DragEvent>;
+  onDragOver?: MuiCancellableEventHandler<React.DragEvent>;
+  onDragEnd?: MuiCancellableEventHandler<React.DragEvent>;
+}
+
+export interface UseTreeItem2ContentSlotPropsFromItemsReordering {
+  onDragEnter?: MuiCancellableEventHandler<React.DragEvent>;
+  onDragOver?: MuiCancellableEventHandler<React.DragEvent>;
+}
+
+export interface UseTreeItem2DragAndDropOverlaySlotPropsFromItemsReordering
+  extends TreeItem2DragAndDropOverlayProps {}
+
+declare module '@mui/x-tree-view/useTreeItem2' {
```

Comment:
One limitation of using `declare` here is that if you use one `RichTreeView` and one `RichTreeViewPro`, and if this `RichTreeView` use a custom Tree Item, then this custom Tree Item will have a drag & drop props on the value returned by `getRootProps` and the other functions.

I think it's fine